### PR TITLE
[14.0][MIG] stock_reserve_sale

### PR DESCRIPTION
**Pull request by mt-software-de**
This is the migration for stock_reserve_sale
there is an other migration for stock_reserve https://github.com/OCA/stock-logistics-warehouse/pull/1346

### CODE REVIEW

**Comment by mt-software-de** [comment]
@rousseldenis maybe you can help me here?
This is the migration for stock_reserve_sale and it depends on https://github.com/OCA/stock-logistics-warehouse/pull/1346
what is here now the right thing to do?

**Comment by rousseldenis** [comment]
> @rousseldenis maybe you can help me here? This is the migration for stock_reserve_sale and it depends on #1346 what is here now the right thing to do?

Could you have a review on #1346 ?

File: stock_reserve_sale/models/sale.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+                    is_stock_reservable = True
+            if sale.state not in ("draft", "sent"):
+                is_stock_reservable = False
+            sale.is_stock_reservable = is_stock_reservable
+            sale.has_stock_reservation = has_stock_reservation
+
+    has_stock_reservation = fields.Boolean(
+        compute="_compute_stock_reservation",
+        readonly=True,
+        multi="stock_reservation",
```
Comment (gdgellatly) [inline stock_reserve_sale/models/sale.py]: `WARNING prod odoo.fields: Field sale.order.has_stock_reservation: unknown parameter 'multi', if this is an actual parameter you may want to override the method _valid_field_parameter on the relevant model in order to allow it`

File: stock_reserve_sale/models/sale.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        store=True,
+        string="Has Stock Reservations",
+    )
+    is_stock_reservable = fields.Boolean(
+        compute="_compute_stock_reservation",
+        readonly=True,
+        multi="stock_reservation",
```
Comment (gdgellatly) [inline stock_reserve_sale/models/sale.py]: `WARNING prod odoo.fields: Field sale.order.is_stock_reservable: unknown parameter 'multi', if this is an actual parameter you may want to override the method _valid_field_parameter on the relevant model in order to allow it`

File: stock_reserve_sale/view/sale.xml [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                expr="//field[@name='order_line']/form//field[@name='state']"
+                position="before"
+            >
+                <field name="reservation_ids" invisible="1" />
+                <button
+                    name="%(action_sale_stock_reserve)d"
+                    type="action"
+                    string="Reserve Stock"
+                    attrs="{'invisible': ['|', ('reservation_ids', '!=', []),
+                                                ('state', '!=', 'draft')]}"
```
Comment (gdgellatly) [inline stock_reserve_sale/view/sale.xml]: I think this should also allow `sent` state

File: stock_reserve_sale/view/sale.xml [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                />
+                <button
+                    name="release_stock_reservation"
+                    type="object"
+                    string="Release Reservation"
+                    attrs="{'invisible': ['|', ('reservation_ids', '=', []),
+                                                ('state', '!=', 'draft')]}"
```
Comment (gdgellatly) [inline stock_reserve_sale/view/sale.xml]: I think this should also allow `sent` state

**Review by gdgellatly — COMMENTED** [review]
I've given a thorough functional review of #1346 and #1347 and there is quite a bit I didn't like.

Firstly the generation of the Reservation location and the defaulting everywhere. It is 100% not multicompany compatible and just leads to constant errors. We shouldn't have a module that requires development out of the box to work MC.

Now in my case I solved it by creating an init hook that creates a Reservation location for each warehouse, added a warehouse_id column with onchanges, and set the picking type from that warehouse. Otherwise in multicompany you get 2 location errors and a picking type error by default. But really I feel like these defaults should just not exist at all.

We can default company and use an onchange there (and if adding a warehouse field, onchanges there for locations and company). In sale_stock_reserve, we can also get the warehouse from the sale.

I also feel like at the point of reserving stock, that perhaps it should try to reserve then, not when the scheduler runs, but not so important.

Next (in #1347) the sale views should allow draft,sent states not just draft. If anything, it is more likely you would reserve stock, after emailing a customer a quote.

Finally in sale for calculating is_stock_reservable of the line. That boolean is ultra confusing and doesn't work. Everything just ends up True that I tested, so I unrolled and just if/elif the whole thing and it works as expected. I think the issue probably comes from the procure method check, because in my customisation we unpack that to reserve the BoM raw materials. - sorry it is both together I reviewed as one will copy to other.

**Review by gdgellatly — COMMENTED** [review]
On existing databases the computed field (I am assuming) may have a very long loading time. This one wasn't even that big and took over 8 minutes.

2022-06-19 11:17:29,188 1 INFO prod odoo.modules.loading: Module stock_reserve_sale loaded in 478.65s, 78893 queries

File: stock_reserve_sale/models/sale.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
+# Copyright 2013 Camptocamp SA - Guewen Baconnier
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import api, fields, models
+from odoo.exceptions import UserError, except_orm
```
Comment (gdgellatly) [inline stock_reserve_sale/models/sale.py]:
```suggestion
from odoo.exceptions import UserError
```

except_orm is a deprecated alias to UserError

File: stock_reserve_sale/models/sale.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+        block_on_reserve = ("product_id", "product_uom", "type")
+        update_on_reserve = ("price_unit", "product_uom_qty")
+        keys = set(vals.keys())
+        test_block = keys.intersection(block_on_reserve)
+        test_update = keys.intersection(update_on_reserve)
+        if test_block:
+            for line in self:
+                if not line.reservation_ids:
+                    continue
+                raise except_orm(
```
Comment (gdgellatly) [inline stock_reserve_sale/models/sale.py]:
```suggestion
                raise UserError(
```

File: stock_reserve_sale/models/sale.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+                        "before changing the product."
+                    ),
+                )
+        res = super().write(vals)
+        if test_update:
+            for line in self:
+                if not line.reservation_ids:
+                    continue
+                if len(line.reservation_ids) > 1:
+                    raise except_orm(
```
Comment (gdgellatly) [inline stock_reserve_sale/models/sale.py]:
```suggestion
                    raise UserError(
```

**Comment by victoralmau** [comment]
Please, cherry-pick https://github.com/OCA/stock-logistics-warehouse/pull/1489

**Comment by rousseldenis** [comment]
@mt-software-de Could you fix pre-commit ?

**Comment by mt-software-de** [comment]
> Please, cherry-pick #1489

done

**Comment by mt-software-de** [comment]
> @mt-software-de Could you fix pre-commit ?

done

**Comment by rousseldenis** [comment]
@mt-software-de You should fix this as you lost all the commit history

**Comment by rvalyi** [comment]
superseded by https://github.com/OCA/stock-logistics-warehouse/pull/1696

**Comment by rvalyi** [comment]
@mt-software-de I took the initiative to redo your PR (see my previous comment) because your last action on this PR was some months ago and I wasn't aware you were still active in the OCA at the moment... Instead as the module was also being ported to v16 I thought it was urgent to get your fixes included... Finally incorporating your work atop of the module history wasn't trivial and I was not sure you would/could do it. But don't worry your work and commit is preserved in the other migration PR I mentioned I my previous comment.

**Comment by mt-software-de** [comment]
> @mt-software-de I took the initiative to redo your PR (see my previous comment) because your last action on this PR was some months ago and I wasn't aware you were still active in the OCA at the moment... Instead as the module was also being ported to v16 I thought it was urgent to get your fixes included... Finally incorporating your work atop of the module history wasn't trivial and I was not sure you would/could do it. But don't worry your work and commit is preserved in the other migration PR I mentioned I my previous comment.

No problem, i'm active in the OCA more than ever. But for me there was currently no focus on this.
Because we didn't used the module anymore, we achieved our needs with an own separate module. 
Thanks for finishing it.

**Comment by rvalyi** [comment]
@mt-software-de actually I'm curious: did you find big issues with the stock_reserve_sale module? What kind of alternative are you using? (you could even send me your module if that's it's not confidential at rvalyi - at - akretion.com). We are actually still testing this module so I would be curious for alternative approches. I was planning to use stock_demand_estimate before, I think there is some overlap, but I discovered stock_reserve_sale recently and thought it might be closer to our need.

**Comment by mt-software-de** [comment]
> @mt-software-de actually I'm curious: did you find big issues with the stock_reserve_sale module? What kind of alternative are you using? (you could even send me your module if that's it's not confidential at rvalyi - at - akretion.com). We are actually still testing this module so I would be curious for alternative approches. I was planning to use stock_demand_estimate before, I think there is some overlap, but I discovered stock_reserve_sale recently and thought it might be closer to our need.

I'll clarify it today to see if i can push it.

**Comment by mt-software-de** [comment]
> @mt-software-de actually I'm curious: did you find big issues with the stock_reserve_sale module? What kind of alternative are you using? (you could even send me your module if that's it's not confidential at rvalyi - at - akretion.com). We are actually still testing this module so I would be curious for alternative approches. I was planning to use stock_demand_estimate before, I think there is some overlap, but I discovered stock_reserve_sale recently and thought it might be closer to our need.

This is our approach https://github.com/OCA/stock-logistics-warehouse/pull/1700

**Pull request closed without merging**